kratos/tensor_adaptors/tensor_adaptor.h

```diff
@@ -0,0 +1,421 @@
+         *
+         */
+        ~Storage();
+
+        ///@}
+        ///@name Public operations
+        ///@{
+
+        /**
+         * @brief Creates a copy of the current storage object.
+         * @details This method returns a pointer to a new Storage object that is having copied data and copied
+         *          pointer to the container.
+         * @return Storage::Pointer A pointer to the copied Storage object.
+         */
+        Storage::Pointer Copy() const;
+
+        /**
+         * @brief Moves the internal data.
+         * @warning The TensorAdaptor should not be used after the move is called.
+         *          The management of the data should be carried out by the owner of the
+         *          returned Kratos::span. Otherwise, there will be memory leaks.
+         * @throws If the internal data is already moved.
+         * @return Kratos::span<TDataType>  Returns a span containing the internal data.
+         */
+        Kratos::span<TDataType> MoveData();
+
+        /**
+         * @brief Return a view of the internal data structure.
+         * @throws If the internal data is already moved.
+         */
+        Kratos::span<const TDataType>  ViewData() const;
+
+        /**
+         * @brief Return a view of the internal data structure.
+         * @throws If the internal data is already moved.
+         */
+        Kratos::span<TDataType> ViewData();
+
+        /**
+         * @brief Returns the shape of the tensor as a dense vector of unsigned integers.
+         * @details This function provides the dimensions of the tensor, where each element in the returned
+         *          DenseVector corresponds to the size of the tensor in that particular dimension. First
+         *          dimension always represents how many entities are in the container.
+         *
+         * @return DenseVector<unsigned int> A vector containing the size of each dimension of the tensor.
+         */
+        DenseVector<unsigned int> Shape() const;
+
+        /**
+         * @brief Returns the shape of the underlying tensor data as a dense vector.
+         * @details This method returns the shape of the data which may be collected and stored
+         *          in each entity of the specified container. This is always one dimension less
+         *          than the @ref Storage::Shape method.
+         * @return DenseVector<unsigned int> A vector containing the dimensions of the tensor.
+         */
+        DenseVector<unsigned int> DataShape() const;
+
+
+        /**
+         * @brief Returns the number of elements in the tensor adaptor.
+         * @return The size as an unsigned integer.
+         */
+        unsigned int Size() const;
+
+        /**
+         * @brief Returns a pointer to the underlying container.
+         * @details This method provides access to the internal container used by the tensor adaptor.
+         *          The returned pointer allows read-only operations on the container.
```

Comment:
This is a lie. The returned pointer points to a mutable instance, so write operations are also allowed. Is the documentation or the implementation wrong?

```suggestion
         *          The returned pointer allows read-only operations on the container.
```

Reply:
Fixed